src/ext/java/org/opentripplanner/ext/siri/ModifiedTripBuilder.java

```diff
@@ -0,0 +1,249 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(TimetableHelper.class);
+  private final TripTimes existingTripTimes;
+  private final TripPattern pattern;
+  private final LocalDate serviceDate;
+  private final ZoneId zoneId;
+  private final EntityResolver entityResolver;
+  private final List<CallWrapper> calls;
+  private final boolean cancellation;
+  private final OccupancyEnumeration occupancy;
+  private final boolean predictionInaccurate;
+
+  public ModifiedTripBuilder(
+    TripTimes existingTripTimes,
+    TripPattern pattern,
+    EstimatedVehicleJourney journey,
+    LocalDate serviceDate,
+    ZoneId zoneId,
+    EntityResolver entityResolver
+  ) {
+    this.existingTripTimes = existingTripTimes;
+    this.pattern = pattern;
+    this.serviceDate = serviceDate;
+    this.zoneId = zoneId;
+    this.entityResolver = entityResolver;
+
+    calls = CallWrapper.of(journey);
+    cancellation = TRUE.equals(journey.isCancellation());
+    predictionInaccurate = TRUE.equals(journey.isPredictionInaccurate());
+    occupancy = journey.getOccupancy();
+  }
+
+  /**
+   * Constructor for tests
+   */
+  public ModifiedTripBuilder(
+    TripTimes existingTripTimes,
+    TripPattern pattern,
+    LocalDate serviceDate,
+    ZoneId zoneId,
+    EntityResolver entityResolver,
+    List<CallWrapper> calls,
+    boolean cancellation,
+    OccupancyEnumeration occupancy,
+    boolean predictionInaccurate
+  ) {
+    this.existingTripTimes = existingTripTimes;
+    this.pattern = pattern;
+    this.serviceDate = serviceDate;
+    this.zoneId = zoneId;
+    this.entityResolver = entityResolver;
+    this.calls = calls;
+    this.cancellation = cancellation;
+    this.occupancy = occupancy;
+    this.predictionInaccurate = predictionInaccurate;
+  }
+
+  /**
+   * Create a new StopPattern and TripTimes for the trip based on the calls, and other fields read
+   * in form the SIRI-ET update.
+   */
+  public Result<TripUpdate, UpdateError> build() {
+    TripTimes newTimes = new TripTimes(existingTripTimes);
+
+    StopPattern stopPattern = createStopPattern(pattern, calls, entityResolver);
+
+    if (cancellation || stopPattern.isAllStopsCancelled()) {
+      LOG.debug("Trip is cancelled");
+      newTimes.cancelTrip();
+      return Result.success(new TripUpdate(pattern.getStopPattern(), newTimes, serviceDate));
+    }
+
+    applyUpdates(newTimes);
+
+    if (pattern.getStopPattern().equals(stopPattern)) {
+      // This is the first update, and StopPattern has not been changed
+      newTimes.setRealTimeState(RealTimeState.UPDATED);
+    } else {
+      // This update modified stopPattern
+      newTimes.setRealTimeState(RealTimeState.MODIFIED);
+    }
+
+    var result = newTimes.validateNonIncreasingTimes();
+    if (result.isFailure()) {
+      var updateError = result.failureValue();
+      LOG.info(
+        "TripTimes are non-increasing after applying SIRI delay propagation - Trip {}. Stop index {}",
```

Comment:
Isn't this a relatively common error and would therefore spam the logs?

Reply:
This behaviour is carried over from existing code